### PR TITLE
Defer subclass methods if superclass has not been analyzed

### LINKED ISSUE #5548: False negative for overriding decorated method in an import cycle
Consider this test:
```
[case testCyclicDecoratir]
import b
[file a.py]
import b
import c

class A(b.B):
    @c.deco
    def meth(self) -> int: ...
[file b.py]
import a
import c

class B:
    @c.deco
    def meth(self) -> int: ...
[file c.py]
from typing import TypeVar, Tuple, Callable
T = TypeVar('T')
def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
[out]
```
It currently fails with
```
tmp/a.py:6: error: Signature of "meth" incompatible with supertype "B"
```
Although it is obviously compatible (it is actually identical). The error goes away if I change the order files are processed in the cycle (by importing `a` in main).

The reason is that non-trivial decorator needs to be type-checked to find the type in superclass _first_. A potential solution is to just defer the subclass method node.

### LINKED ISSUE #5560: mypy error trying to check pymc3
* Are you reporting a bug, or opening a feature request?
Bug report, for the following error:
    pymc3/backends/hdf5.py:101: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues version: 0.620
* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.
Tried testing the code in this repo: 
origin	git@github.com:rpgoldman/pymc3.git
* What is the actual behavior/output?
See error report above.
* What is the behavior/output you expect?
No runtime error.
* What are the versions of mypy and Python you are using?
mypy 0.620
* Do you see the same issue after installing mypy from Git master?
Will check and then update this ticket
* What are the mypy flags you are using? (For example --strict-optional)
    mypy --show-traceback --ignore-missing-imports pymc3/ > /tmp/mypy-transcript.txt
* If mypy crashed with a traceback, please paste
  the full traceback below.

````
pymc3/model.py:672: error: Decorated property not supported
pymc3/variational/opvi.py:1556: error: Decorated property not supported
pymc3/variational/flows.py:193: error: Decorated property not supported
pymc3/__init__.py:25: error: Name 'utils' already defined (by an import)
pymc3/data.py:226: error: Need type annotation for 'RNG'
Traceback (most recent call last):
  File "/Users/rpg/Library/Python/3.7/bin/mypy", line 11, in <module>
    sys.exit(console_entry())
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/__main__.py", line 7, in console_entry
    main(None)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/main.py", line 91, in main
    res = type_check_only(sources, bin_dir, options, flush_errors, fscache)  # noqa
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/main.py", line 148, in type_check_only
    fscache=fscache)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/build.py", line 177, in build
    flush_errors, fscache)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/build.py", line 350, in _build
    graph = dispatch(sources, manager)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/build.py", line 2560, in dispatch
    process_graph(graph, manager)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/build.py", line 2853, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/build.py", line 2976, in process_stale_scc
    graph[id].type_check_first_pass()
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/build.py", line 2151, in type_check_first_pass
    self.type_checker().check_first_pass()
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 253, in check_first_pass
    self.accept(d)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 352, in accept
    stmt.accept(self)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/nodes.py", line 808, in accept
    return visitor.visit_class_def(self)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 1368, in visit_class_def
    self.accept(defn.defs)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 352, in accept
    stmt.accept(self)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/nodes.py", line 873, in accept
    return visitor.visit_block(self)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 1514, in visit_block
    self.accept(s)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 352, in accept
    stmt.accept(self)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/nodes.py", line 443, in accept
    return visitor.visit_overloaded_func_def(self)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 385, in visit_overloaded_func_def
    self._visit_overloaded_func_def(defn)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 409, in _visit_overloaded_func_def
    self.check_method_override(defn)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 1198, in check_method_override
    self.check_method_or_accessor_override_for_base(defn, base)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 1208, in check_method_or_accessor_override_for_base
    self.check_method_override_for_base_with_name(defn, name, base)
  File "/Users/rpg/Library/Python/3.7/lib/python/site-packages/mypy/checker.py", line 1252, in check_method_override_for_base_with_name
    assert False, str(base_attr.node)
AssertionError: Var(sampler_vars)
````
(You can freely edit this text, please remove all the lines
you believe are unnecessary.)

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #5560 
Fixes #5548 

Half of the PR is updating various function signatures to also accept `Decorator`. I still prohibit `Decorator` as a target in fine-grained mode, but I think there should be no harm to defer it in normal mode.

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -338,20 +336,23 @@ def check_top_level(self, node: MypyFile) -> None:
         assert not self.current_node_deferred
         # TODO: Handle __all__
 
+    def defer_node(self, node: DeferredNodeType, enclosing_class: Optional[TypeInfo]) -> None:
+        if self.errors.type_name:
+            type_name = self.errors.type_name[-1]
+        else:
+            type_name = None
+        # Shouldn't we freeze the entire scope?
```
[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply mypy/checker.py]: I am not sure what was the intent. This comment is two years old, maybe just remove it?

File: mypy/server/update.py [hunk 2]
```diff
@@ -1058,7 +1058,7 @@ def is_verbose(manager: BuildManager) -> bool:
 
 
 def target_from_node(module: str,
-                     node: Union[FuncDef, MypyFile, OverloadedFuncDef, LambdaExpr]
+                     node: Union[FuncDef, MypyFile, OverloadedFuncDef, LambdaExpr, Decorator]
```
[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply mypy/server/update.py]: I can remove `Decorator` from all these unions, but this will require a couple of asserts.

[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/server/update.py]: I think that it's better to have the caller ensure that the function is called with the correct types. The annotation as it is now is arguably misleading.

File: mypy/checker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/checker.py]: Maybe reword it like this: "We don't freeze the entire scope since only top-level functions and methods can be deferred. Only module/class level scope information is needed. Module-level scope information is preserved in the TypeChecker instance." (I *think* that this is correct.)

File: mypy/server/update.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/server/update.py]: Ah, good point! In that case maybe `LambdaExpr` should also be removed from the annotation?

[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply mypy/server/update.py]: Yes, I will remove both.

**Comment by ilevkivskyi** [comment]
@JukkaL Thanks for review! I addressed the comments. I decided to have a separate `FineDeferredNode` named tuple (since named tuples can't be generic). This way we can avoid all asserts and extra items in unions in arguments.

I also added a bunch of tests combining overloads, decorators, variables, double deferrals, and super calls (also with various combinations of these).

File: mypy/checker.py [hunk 3]
```diff
@@ -1253,15 +1276,26 @@ def expand_typevars(self, defn: FuncItem,
         else:
             return [(defn, typ)]
 
-    def check_method_override(self, defn: Union[FuncBase, Decorator]) -> None:
-        """Check if function definition is compatible with base classes."""
+    def check_method_override(self, defn: Union[FuncDef, OverloadedFuncDef, Decorator]) -> None:
+        """Check if function definition is compatible with base classes.
+
+        This may defer the method if a signature is not available in at least on base class.
```
Comment (JukkaL) [inline mypy/checker.py]: Typo: "on" -> "one"

File: test-data/unit/check-classes.test [hunk 4]
```diff
@@ -5117,3 +5117,295 @@ class C:
...
+    def __init__(self) -> None:
+        self.x = f()
+def f() -> int: ...
+[file c.py]
+from typing import TypeVar, Tuple, Callable
+T = TypeVar('T')
+def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
+[out]
+
+[case testCyclicDecoratorSuperDeferred]
```
Comment (JukkaL) [inline test-data/unit/check-classes.test]: Did you verify that these test cases trigger deferral? Using `from ... import` can be used to make it more explicit in which order modules are processed by mypy.

File: mypy/checker.py [hunk 5]
```diff
@@ -72,19 +73,31 @@
 
 DEFAULT_LAST_PASS = 1  # type: Final  # Pass numbers start at 0
 
+DeferredNodeType = Union[FuncDef, LambdaExpr, OverloadedFuncDef, Decorator]
+FineDeferredNodeType = Union[FuncDef, MypyFile, OverloadedFuncDef]
```
Comment (JukkaL) [inline mypy/checker.py]: Style nit: I'd prefer to call this `FineGrainedDeferredNodeType` or something. 

We don't use "fine" as a shorted version of "fine-grained" elsewhere and it can be confusing.

File: mypy/checker.py [hunk 6]
```diff
@@ -72,19 +73,31 @@
...
-        ('node', Union[FuncDef, LambdaExpr, MypyFile, OverloadedFuncDef]),
+        ('node', DeferredNodeType),
         ('context_type_name', Optional[str]),  # Name of the surrounding class (for error messages)
         ('active_typeinfo', Optional[TypeInfo]),  # And its TypeInfo (for semantic analysis
                                                   # self type handling)
     ])
 
+# Same as above, but for fine-grained mode targets. Only top-level functions/methods
+# and module top levels are allowed as such.
+FineDeferredNode = NamedTuple(
```
Comment (JukkaL) [inline mypy/checker.py]: Similar to above, I'd prefer to name this `FineGrainedDeferredNode`.

File: test-data/unit/check-classes.test (hunk 4 above)
Comment (ilevkivskyi) [reply test-data/unit/check-classes.test]: I actually checked only first two tests I added, I will also double-check rest just in case.

Comment (ilevkivskyi) [reply test-data/unit/check-classes.test]: They all get deferred expected number of times (1, 2, or 3), I switched to `from ... import` in half of the tests anyway.
